### DISCUSSION TITLE
Will the structure of custom field.value be supported?

### DISCUSSION THREAD

**Discussion by tuluffy** [react-hook-form/react-hook-form#5121] [Q&A] [answered]
**I used a third-party component library, which contains two components: Select and Input. In the Select-Component, field.value is a Number, and in the Input-Component, field.value is an Object (example: {value, ErrorMessage, event}), as shown in the figure.**

![image](https://user-images.githubusercontent.com/26808339/117641530-d7f21f00-b1b8-11eb-8529-8fdd3087ff7e.png)

![image](https://user-images.githubusercontent.com/26808339/117641641-fb1cce80-b1b8-11eb-920e-43938600c21a.png)

**Is it possible to customize the structure of field.value instead of the default structure provided by the component?**

**Comment by thanh-nguyen-95** [comment, marked as answer]
If I understand it correctly, what you're trying to do is having custom `value` and `onChange` behaviour in third-party component and reflect it on the form's values.

Here's a code example of how I would do it:

```jsx
// for the input
<Controller 
  name="end"
  rules={{
    // you have to use `validate` here because your value is 'customized'
    validate: (valueInfo) => {
      const isFieldValid = !!valueInfo.value

      if (!isFieldValid) {
        // return error message if it's invalid
        return 'End is required'
      }
      // return `true` otherwise
      return true
    }
  }}
  defaultValue={{ value: 'test', errorMessage: '', ev: null }}
  render={({ field, fieldState }) => {
    const { value: valueInfo } = field

    return (
      <Input
        {...field}
        value={valueInfo.value}
        onChange={(ev) => {
          field.onChange({
            value: ev.target.value,
            errorMessage: fieldState.error?.message,
            event: ev
          })
        }}
      />
    )
  }}
/>
```

Key takeaways here:

- You must use `validate` to validate your value. I don't think pre-defined ones can handle custom value
- Provide consistent and correct `defaultValue`
- Override `value` and `onChange` when rendering fields

Since I don't know your intention or the library that you're using, I can't give you a clear answer. If it's not your intention at all, please provide a more detail use case or a CodeSandBox (CSB).

TL;DR: I believe it's possible but I don't recommend doing it this way

**Reply by bluebill1049** [reply]
thank you @thanh-nguyen-95 ❤️
